libfastpm/neutrinos_lra.c

```diff
@@ -0,0 +1,722 @@
+        get_delta_nu_combined(CP, &delta_tot_table, Time, delta_tot_table.delta_nu_last);
+        /* Decide whether we save the current time or not */
+        if (Time > exp(delta_tot_table.scalefact[delta_tot_table.ia-2]) + 0.009) {
+            /* If so update delta_tot(a) correctly, overwriting current power spectrum */
+            update_delta_tot(&delta_tot_table, Time, Power_in, delta_tot_table.delta_nu_last, 1);
+        }
+        /*Otherwise discard the last powerspectrum*/
+        else
+            delta_tot_table.ia--;
+
+        fastpm_info("Done getting neutrino power: nk = %d, k = %g, delta_nu = %g, delta_cdm = %g,\n", delta_tot_table.nk, delta_tot_table.wavenum[1], delta_tot_table.delta_nu_last[1], Power_in[1]);
+        /*kspace_prefac = M_nu (analytic) / M_particles */
+        const double OmegaNu = get_omega_nu(Time, delta_tot_table.cosmo);
+        /* Omega0 - Omega in neutrinos = Omega in particles*/
+        nupow->nu_prefac = OmegaNu/(delta_tot_table.Omeganonu/pow(Time,3));
+    }
+    double * delta_nu_ratio = (double *) malloc(delta_tot_table.nk * sizeof(double));
+    double * logwavenum = (double *) malloc(delta_tot_table.nk * sizeof(double));
+    gsl_interp * pkint = gsl_interp_alloc(gsl_interp_linear, delta_tot_table.nk);
+    gsl_interp_accel * pkacc = gsl_interp_accel_alloc();
+    /*We want to interpolate in log space*/
+    for(i=0; i < delta_tot_table.nk; i++) {
+        /*Enforce positivity for sanity reasons*/
+        if(delta_tot_table.delta_nu_last[i] < 0)
+            delta_tot_table.delta_nu_last[i] = 0;
+        delta_nu_ratio[i] = delta_tot_table.delta_nu_last[i]/ Power_in[i];
+        logwavenum[i] = log(delta_tot_table.wavenum[i]);
+    }
+    if(delta_tot_table.nk != ps->size)
+        free(Power_in);
+    gsl_interp_init(pkint, logwavenum, delta_nu_ratio, delta_tot_table.nk);
+
+    /*We want to interpolate in log space*/
+    for(i=0; i < ps->size; i++) {
+        if(ps->size == delta_tot_table.nk)
+            nupow->delta_nu_ratio[i] = delta_nu_ratio[i];
+        else {
+            double logk = nupow->logknu[i];
+            if(logk > pkint->xmax)
+                logk = pkint->xmax;
+            nupow->delta_nu_ratio[i] = gsl_interp_eval(pkint, logwavenum, delta_nu_ratio, logk, pkacc);
+        }
+    }
+
+    gsl_interp_accel_free(pkacc);
+    gsl_interp_free(pkint);
+    free(logwavenum);
+    free(delta_nu_ratio);
+    // fastpm_info("Neutrino ratio: nk = %d, k = %g, delta_nu = %g,\n", ps->size, nupow->logknu[1], nupow->delta_nu_ratio[1]);
+}
+
+/*Save the neutrino power spectrum to a file*/
+
+void powerspectrum_nu_save(FastPMPowerSpectrum * ps, char powerspectrum_file[], double MtotbyMcdm)
+{
+    /* substitute the last neutrino power spectrum */
+    int i;
+    /* The k bins are the same because this is constructed from the same power spectrum*/
+    for(i = 0; i < ps->base.size; i++)
+        ps->base.f[i] = pow(delta_tot_table.delta_nu_last[i]/MtotbyMcdm, 2);
```

Comment:
What is the need for the `MtotbyMcdm` scaling? Is `delta_nu_last` not really $\delta_\nu$?

Reply:
In MP-Gadget all the power spectra are delta_X * M_X and so dimensionful. delta_nu inherits the same units that delta_c has. If fastpm really computes delta_c then the MtotbyMcdm factor is not needed.